jans-cedarling/cedarling/src/jwt/decoding_strategy.rs

```diff
@@ -5,140 +5,197 @@
+    ///
+    /// # Errors
+    /// Returns an error if the specified algorithm is unrecognized or the key service initialization fails.
+    pub fn new_with_validation(
+        config_algs: Vec<jwt::Algorithm>,
+        trusted_idps: Vec<TrustedIssuer>,
+    ) -> Result<Self, Error> {
+        // initialize the key service with OpenID configuration endpoints
+        let openid_conf_endpoints = trusted_idps
+            .iter()
+            .map(|x| x.openid_configuration_endpoint.as_ref())
+            .collect();
+        let key_service = KeyService::new(openid_conf_endpoints).map_err(Error::KeyServiceError)?;
+
+        Ok(Self::WithValidation {
+            key_service,
+            supported_algs: config_algs,
+        })
+    }
+
+    /// Decodes a JWT token according to the current decoding strategy.
+    ///
+    /// # Arguments
+    /// - `jwt` - The JWT string to decode.
+    /// - `iss` - Optional expected issuer for validation.
+    /// - `aud` - Optional expected audience for validation.
+    /// - `sub` - Optional expected subject for validation.
+    /// - `validate_nbf`: A boolean indicating whether to validate the "not before" claim.
+    /// - `validate_exp`: A boolean indicating whether to validate the expiration claim.
+    ///
+    /// # Errors
+    /// Returns an error if decoding or validation fails.
+    pub fn decode<T: DeserializeOwned>(
         &self,
         jwt: &str,
         iss: Option<impl ToString>,
         aud: Option<impl ToString>,
-        req_sub: bool,
+        sub: Option<impl ToString>,
+        validate_nbf: bool,
+        validate_exp: bool,
     ) -> Result<T, Error> {
         match self {
             DecodingStrategy::WithoutValidation => self.extract_claims(jwt),
             DecodingStrategy::WithValidation {
                 key_service,
                 supported_algs,
-            } => decode_and_validate_jwt(jwt, iss, aud, req_sub, supported_algs, key_service),
+            } => decode_and_validate_jwt(
+                jwt,
+                iss,
+                aud,
+                sub,
+                validate_nbf,
+                validate_exp,
+                supported_algs,
+                key_service,
+            ),
         }
     }
-}
 
-/// Trait implementation for extracting claims without validation.
-impl ExtractClaims for DecodingStrategy {
-    /// Extracts the claims from a JWT token without validation.
-    fn extract_claims<T: DeserializeOwned>(&self, jwt_str: &str) -> Result<T, Error> {
+    /// Extracts the claims from a JWT token without performing validation.
+    ///
+    /// This method uses a default insecure validator that skips signature
+    /// validation and other checks (e.g., expiration). Only use in trusted environments.
+    ///
+    /// # Errors
+    /// Returns an error if the claims cannot be extracted.
+    pub fn extract_claims<T: DeserializeOwned>(&self, jwt_str: &str) -> Result<T, Error> {
```

Comment:
Looks like we can extract this method to function. `&self` is unused.

Reply:
it's better for organization so i can call it like this in `JwtService::decode_tokens`:
```rust
// extract claims without validation
let access_token_claims = self.decoding_strategy.extract_claims(access_token)?;
let id_token_claims = self.decoding_strategy.extract_claims(id_token)?;
let userinfo_token_claims = self.decoding_strategy.extract_claims(userinfo_token)?;
```

Reply:
i don't want to call `extract_claims` outside the context of `decoding_strategy`

Reply:
found a nice middle ground by doing this [b1a5d10](https://github.com/JanssenProject/jans/pull/9939/commits/b1a5d10c0e902a0cbd9bc3e6391b7a5a2d513d61). This keeps it in the `impl` block while avoiding the unnecessary use of `self`.